File: primitives/hash_test.go
package primitives

import (
	"testing"
)

func TestHashFromHex(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		wantErr bool
	}{
		{
			name:    "valid hash",
			input:   "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			wantErr: false,
		},
		{
			name:    "zero hash",
			input:   "0x0000000000000000000000000000000000000000000000000000000000000000",
			wantErr: false,
		},
		{
			name:    "invalid length (too short)",
			input:   "0x1234567890abcdef",
			wantErr: true,
		},
		{
			name:    "invalid length (too long)",
			input:   "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdefff",
			wantErr: true,
		},
		{
			name:    "missing prefix",
			input:   "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			wantErr: true,
		},
		{
			name:    "invalid character",
			input:   "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdeg",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			hash, err := HashFromHex(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("HashFromHex(%q) error = %v, wantErr %v", tt.input, err, tt.wantErr)
				return
			}
			if !tt.wantErr && hash.IsZero() && tt.input != "0x0000000000000000000000000000000000000000000000000000000000000000" {
				t.Errorf("HashFromHex(%q) returned zero hash unexpectedly", tt.input)
			}
		})
	}
}

func TestHashFromBytes(t *testing.T) {
	tests := []struct {
		name    string
		input   []byte
		wantErr bool
	}{
		{
			name:    "valid 32 bytes",
			input:   make([]byte, 32),
			wantErr: false,
		},
		{
			name:    "too short",
			input:   make([]byte, 31),
			wantErr: true,
		},
		{
			name:    "too long",
			input:   make([]byte, 33),
			wantErr: true,
		},
		{
			name:    "empty",
			input:   []byte{},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := HashFromBytes(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("HashFromBytes() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestHashToHex(t *testing.T) {
	input := "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
	hash, err := HashFromHex(input)
	if err != nil {
		t.Fatalf("HashFromHex() failed: %v", err)
	}

	result := hash.ToHex()
	if result != input {
		t.Errorf("ToHex() = %s, want %s", result, input)
	}
}

func TestHashIsZero(t *testing.T) {
	zeroHash := Hash{}
	if !zeroHash.IsZero() {
		t.Error("Zero hash should return true for IsZero()")
	}

	nonZeroHash, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
	if nonZeroHash.IsZero() {
		t.Error("Non-zero hash should return false for IsZero()")
	}

	// Test ZeroHash constant
	if !ZeroHash.IsZero() {
		t.Error("ZeroHash constant should be zero")
	}
}

func TestHashEquals(t *testing.T) {
	hash1, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
	hash2, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
	hash3, _ := HashFromHex("0xfedcba0987654321fedcba0987654321fedcba0987654321fedcba0987654321")

	if !hash1.Equals(hash2) {
		t.Error("Identical hashes should be equal")
	}

	if hash1.Equals(hash3) {
		t.Error("Different hashes should not be equal")
	}
}

func TestHashEqualsConstantTime(t *testing.T) {
	// This test verifies that Equals uses constant-time comparison
	// We can't directly test timing, but we verify it uses the right function
	hash1, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
	hash2, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")

	// Should be equal
	if !hash1.Equals(hash2) {
		t.Error("Equal hashes should return true")
	}

	// Modify one byte
	hash2[0] = 0xff
	if hash1.Equals(hash2) {
		t.Error("Different hashes should return false")
	}
}

func TestHashBytes(t *testing.T) {
	hash, _ := HashFromHex("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
	bytes := hash.Bytes()

	if len(bytes) != 32 {
		t.Errorf("Bytes() length = %d, want 32", len(bytes))
	}

	// Verify first few bytes
	expected := []byte{0x12, 0x34, 0x56, 0x78}
	for i := 0; i < 4; i++ {
		if bytes[i] != expected[i] {
			t.Errorf("Bytes()[%d] = 0x%02x, want 0x%02x", i, bytes[i], expected[i])
		}
	}
}

func TestHashString(t *testing.T) {
	input := "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
	hash, _ := HashFromHex(input)
	str := hash.String()

	if str != input {
		t.Errorf("String() = %s, want %s", str, input)
	}
}

func TestZeroHashConstant(t *testing.T) {
	expected := "0x0000000000000000000000000000000000000000000000000000000000000000"
	result := ZeroHash.ToHex()

	if result != expected {
		t.Errorf("ZeroHash.ToHex() = %s, want %s", result, expected)
	}
}

func TestHashRoundTrip(t *testing.T) {
	original := "0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef"
	hash, err := HashFromHex(original)
	if err != nil {
		t.Fatalf("HashFromHex() failed: %v", err)
	}

	// Convert to bytes and back
	bytes := hash.Bytes()
	hash2, err := HashFromBytes(bytes)
	if err != nil {
		t.Fatalf("HashFromBytes() failed: %v", err)
	}

	if !hash.Equals(hash2) {
		t.Error("Round trip through bytes failed")
	}

	// Convert to hex and back
	hex := hash.ToHex()
	hash3, err := HashFromHex(hex)
	if err != nil {
		t.Fatalf("HashFromHex() failed: %v", err)
	}

	if !hash.Equals(hash3) {
		t.Error("Round trip through hex failed")
	}
}

File: src/primitives/v2/hash.test.ts
import { describe, it, expect } from "vitest"
import * as Hash from "./hash"

describe("Hash", () => {
	describe("fromHex", () => {
		it("creates hash from hex string with 0x prefix", () => {
			const hex =
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
			const hash = Hash.fromHex(hex)
			expect(hash.length).toBe(32)
			expect(hash[0]).toBe(0x12)
			expect(hash[1]).toBe(0x34)
			expect(hash[31]).toBe(0xef)
		})

		it("creates hash from hex string without 0x prefix", () => {
			const hex =
				"1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
			const hash = Hash.fromHex(hex)
			expect(hash.length).toBe(32)
			expect(hash[0]).toBe(0x12)
		})

		it("throws on invalid hex length", () => {
			expect(() => Hash.fromHex("0x1234")).toThrow(
				"Hash hex must be 64 characters",
			)
		})

		it("throws on invalid hex characters", () => {
			expect(() =>
				Hash.fromHex(
					"0xzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz",
				),
			).toThrow("Invalid hex string")
		})
	})

	describe("fromBytes", () => {
		it("creates hash from 32-byte array", () => {
			const bytes = new Uint8Array(32)
			bytes[0] = 0xaa
			bytes[31] = 0xbb
			const hash = Hash.fromBytes(bytes)
			expect(hash.length).toBe(32)
			expect(hash[0]).toBe(0xaa)
			expect(hash[31]).toBe(0xbb)
		})

		it("throws on invalid byte length", () => {
			expect(() => Hash.fromBytes(new Uint8Array(20))).toThrow(
				"Hash must be 32 bytes",
			)
		})

		it("creates copy of input bytes", () => {
			const bytes = new Uint8Array(32)
			const hash = Hash.fromBytes(bytes)
			bytes[0] = 0xff
			expect(hash[0]).toBe(0)
		})
	})

	describe("toHex", () => {
		it("converts hash to hex string", () => {
			const hash = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			const hex = Hash.toHex(hash)
			expect(hex).toBe(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
		})

		it("pads zeros correctly", () => {
			const bytes = new Uint8Array(32)
			bytes[0] = 0x01
			bytes[1] = 0x0a
			const hash = Hash.fromBytes(bytes)
			const hex = Hash.toHex(hash)
			expect(hex.slice(0, 8)).toBe("0x010a00")
		})
	})

	describe("equals", () => {
		it("returns true for equal hashes", () => {
			const a = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			const b = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			expect(Hash.equals(a, b)).toBe(true)
		})

		it("returns false for different hashes", () => {
			const a = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			const b = Hash.fromHex(
				"0x0000000000000000000000000000000000000000000000000000000000000000",
			)
			expect(Hash.equals(a, b)).toBe(false)
		})

		it("uses constant time comparison", () => {
			const a = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			const b = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdee",
			)
			expect(Hash.equals(a, b)).toBe(false)
		})
	})

	describe("Hash factory", () => {
		it("creates hash from hex string", () => {
			const hash = Hash.Hash(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			expect(hash.length).toBe(32)
			expect(hash[0]).toBe(0x12)
		})
	})

	describe("isHash", () => {
		it("returns true for valid hash", () => {
			const hash = Hash.fromHex(
				"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
			)
			expect(Hash.isHash(hash)).toBe(true)
		})

		it("returns false for non-hash", () => {
			expect(Hash.isHash("not a hash")).toBe(false)
			expect(Hash.isHash(new Uint8Array(20))).toBe(false)
			expect(Hash.isHash(null)).toBe(false)
		})
	})

	describe("ZERO_HASH", () => {
		it("is 32 bytes of zeros", () => {
			expect(Hash.ZERO_HASH.length).toBe(32)
			expect(Hash.ZERO_HASH.every((b) => b === 0)).toBe(true)
		})
	})

	describe("keccak256", () => {
		it("throws NotImplementedError", () => {
			expect(() => Hash.keccak256(new Uint8Array(0))).toThrow(
				Hash.NotImplementedError,
			)
			expect(() => Hash.keccak256(new Uint8Array(0))).toThrow(
				"keccak256 not yet implemented",
			)
		})
	})
})
